### ISSUE TITLE
New feature: versions of JSON structures

### ISSUE THREAD

**Issue by bwbeach** [Backblaze/b2-sdk-java#52]
I have an application (the B2 service itself) with multiple API versions.  The differences between the versions are simple that some fields in JSON in and JSON out only exist in some version numbers.  It would be really nice if B2Json could handle that, and drop some fields during serialization if those fields are not in the version requested.

**Comment by bwbeach** [comment]
These are my goals:

- During serialization, drop fields that are not in the version chosen.
- During de-serialization, ignore/disallow fields that are not in the version chosen.

Overall, if the field is not in the version chosen, treat it as if it were not in the structure.

My proposal is to: (1) add new annotations for fields, (2) pass options to calls that say what version to use.

The new annotations would be:

```
@firstVersion(startVersion = s)    // says that a field exists in all versions: s <= v
@versionRange(startVersion = s, endVersion = e)   // field exists in all versions: s <= v && v <= e
```

To tell B2Json what version to use, I'd like to add an Options object as an added parameter to all calls.  Options will default to using version 1 and not allowing extra fields when deserializing.  It will have a builder to change both of those settings.  Existing methods that take `int options` will still be there, but will be deprecated and will call the new method using Options objects.

**Comment by bwbeach** [comment]
I think the plan for the constructor should be:

The constructor gets values for *all* fields as parameters.  It should also get a special parameter that is the version number:

```
@constructor(params = "a, b, c", versionParam = "v")
```

**Comment by bwbeach** [comment]
Updated proposal for version annotations:

```
@firstVersion(firstVersion = s)    // says that a field exists in all versions: s <= v
@versionRange(firstVersion = s, lastVersion = e)   // field exists in all versions: s <= v && v <= e
```

**Comment by bwbeach** [comment]
I remembered that reflection does not provide parameter names for constructors, so there's no need to say what the param name of the version is.  But we do need to know where it is in the param list.

So, the updated proposal is to include the version parameter name in both params and versionParam, like this:

```
@constructor(params = "a, b, c, v", versionParam = "v")
```

**Comment by bwbeach** [comment]
Next question: what to do with required fields when they aren't in the version.

I think the constructor should be given the default value for that type (null, or 0), and the code in the constructor can figure out what to do based on the version number.

**Comment by bwbeach** [comment]
FIxed by #53
